horovod/common/operations.cc

```diff
@@ -1667,64 +575,49 @@ bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx,
         response_list, state.tensor_table, tensor_names);
   }
 
-  if (state.response_cache.capacity() > 0) {
-    std::lock_guard<std::mutex> guard(horovod_global.mutex);
-    // All workers add supported responses to cache. This updates the cache
-    // order consistently across workers.
-    for (auto& response : response_list.responses()) {
-      if (response.response_type() == Response::ResponseType::ALLREDUCE &&
-          (int)response.devices().size() == state.size) {
-        state.response_cache.put(response, state.tensor_table);
-      }
-    }
-
-    // Reassign cache bits based on current cache order.
-    state.response_cache.update_cache_bits();
-  }
-
   // Perform the collective operation. All nodes should end up performing
   // the same operation.
+  int rank = state.controller->GetRank();
   for (auto& response : response_list.responses()) {
-    LOG(TRACE, state.rank) << "Performing " << response.tensor_names_string();
-    LOG(DEBUG, state.rank) << "Processing " << response.tensor_names().size()
+    LOG(TRACE, rank) << "Performing " << response.tensor_names_string();
+    LOG(DEBUG, rank) << "Processing " << response.tensor_names().size()
                            << " tensors";
     PerformOperation(state.tensor_table, response);
-    LOG(TRACE, state.rank) << "Finished performing "
+    LOG(TRACE, rank) << "Finished performing "
                            << response.tensor_names_string();
   }
 
   if (state.param_manager.IsAutoTuning()) {
-    state.param_manager.Update(tensor_names, total_tensor_size);
-  }
+    bool should_sync = state.param_manager.Update(tensor_names,
+                                                  total_tensor_size);
 
-  if (response_list.shutdown()) {
-    should_shut_down = true;
+    if (should_sync){
+      state.controller->SynchronizeParameters(state.param_manager);
+    }
```

Comment:
Need to `state.param_manager.Reset()` after syncing the parameters.

Reply:
Yes. `Reset()` is called inside `SynchronizeParameters`.

Reply:
Ah, I missed that.  Nice!